conbench/api/runs.py

```diff
@@ -386,110 +324,129 @@ def get(self, run_id):
+                loop with an ID that is alphabetically earlier than the cursor value, it
+                will not be included in the next page. This is a known limitation of the
+                current implementation, and it is not expected to be a problem in
+                practice. This is because the number of runs per commit hash is expected
+                to be much less than the page size, and runs are added quite
+                infrequently.
+          - in: query
+            name: page_size
             schema:
               type: integer
+              minimum: 1
+              maximum: 1000
+            description: |
+                The size of pages for pagination (see `cursor`). Default 100. Max 1000.
         tags:
           - Runs
         """
-        sha_arg: Optional[str] = f.request.args.get("sha")
-        commit_hashes = sha_arg.split(",") if sha_arg else None
-
-        days_arg: Optional[int] = f.request.args.get("days")
-        days = int(days_arg) if days_arg else 14
-        if days > 30:
-            self.abort_400_bad_request("days must be no more than 30")
-
-        return [
-            self.serializer.one._dump(run.earliest_result, get_baseline_runs=False)
-            for run in get_all_run_info(
-                min_time=datetime.datetime.now(datetime.timezone.utc)
-                - datetime.timedelta(days=days),
-                max_time=datetime.datetime.now(datetime.timezone.utc),
-                commit_hashes=commit_hashes,
+        filters = []
+
+        commit_hash_arg: Optional[str] = f.request.args.get("commit_hash")
+        if not commit_hash_arg:
+            self.abort_400_bad_request("commit_hash is required")
+        filters.append(Commit.sha.in_(commit_hash_arg.split(",")))
+
+        cursor_arg: Optional[str] = f.request.args.get("cursor")
+        if cursor_arg and cursor_arg != "null":
+            filters.append(BenchmarkResult.run_id > cursor_arg)
+
+        page_size_arg = f.request.args.get("page_size", 100)
+        try:
+            page_size = int(page_size_arg)
+            assert 1 <= page_size <= 1000
+        except Exception:
+            self.abort_400_bad_request(
+                "page_size must be a positive integer no greater than 1000"
+            )
+
+        query = s.select(BenchmarkResult).join(Commit).where(*filters)
+
+        # Aggregate all benchmark results into runs (this is keyed by run_id)
+        run_info: Dict[str, RunAggregate] = {}
+        for result in current_session.scalars(query).all():
+            run_id = result.run_id
+            if run_id in run_info:
+                run_info[run_id].update(result)
+            else:
+                run_info[run_id] = RunAggregate(
+                    earliest_result=result,
+                    result_count=1,
+                    any_result_failed=result.is_failed,
+                )
+
+        this_page_run_ids = sorted(run_info.keys())[:page_size]
```

Comment:
Yeah, so we only limit the page size after executing the full query. Maybe this is dumb and we should never limit the size? or else clients will ask the server to make the same query over and over in those rare cases that page size is less than the total number of runs.

Or I could use [this skip query](https://github.com/conbench/conbench/blob/1f141414edefc721408e723cb5d978ad8e88dfe0/conbench/entities/benchmark_result.py#L942), modified for this endpoint, but it really seems like overkill

Reply:
Actually it'll be a slightly smaller query each time since it's filtered on the cursor, but still redundant.

Reply:
This morning I tested this code against the Arrow data. The query is actually extremely fast, but then the server takes 5 seconds to deaggregate by going through the loop.

I think after all I'll refactor this to use the skip scan query, which should remain quick and require no postprocessing. This should speed this up and also enable querying only `page_size` runs at once.

Reply:
Never mind, didn't need to use the skip scan query. We can use the much simpler DISTINCT ON idea, because we don't need to order by benchmark result timestamp, and we don't need to worry about the results scaling (as JP discussed in https://github.com/conbench/conbench/issues/1466#issuecomment-1711571921).